### PR TITLE
Edition custom property arrays

### PR DESCRIPTION
**Pull request by RicardRC**
Associated proposal: https://github.com/GodotECS/godex/discussions/236

### CODE REVIEW

File: ecs.h [hunk 1]
```diff
@@ -43,6 +43,9 @@ struct DataAccessorFuncs {
 	bool (*set_by_index)(void *p_self, const uint32_t p_parameter_index, const Variant &p_data) = nullptr;
 	bool (*get_by_index)(const void *p_self, const uint32_t p_parameter_index, Variant &r_data) = nullptr;
 	void (*call)(void *p_self, const StringName &p_method, const Variant **p_args, int p_argcount, Variant *r_ret, Callable::CallError &r_error) = nullptr;
+	bool (*dynamic_get)(const void *p_self, const StringName &p_name, const Variant &r_property) = nullptr;
+	bool (*dynamic_set)(void *p_self, const StringName &p_name, Variant &p_property) = nullptr;
+	bool (*dynamic_get_property_list)(const void *p_self, List<PropertyInfo> *r_list) = nullptr;
```
Comment (AndreaCatania) [inline ecs.h:48]: This should return `void`

File: ecs.h [hunk 2]
```diff
@@ -250,6 +253,9 @@ class ECS : public Object {
 	static bool unsafe_component_set_by_index(godex::component_id p_component_id, void *p_component, uint32_t p_index, const Variant &p_data);
 	static bool unsafe_component_get_by_index(godex::component_id p_component_id, const void *p_component, uint32_t p_index, Variant &r_data);
 	static void unsafe_component_call(godex::component_id p_component_id, void *p_component, const StringName &p_method, const Variant **p_args, int p_argcount, Variant *r_ret, Callable::CallError &r_error);
+	static bool unsafe_component_dynamic_get(godex::component_id p_component_id, const void *p_component, const StringName &p_name, const Variant &p_property);
+	static bool unsafe_component_dynamic_set(godex::component_id p_component_id, const void *p_component, const StringName &p_name, Variant &r_property);
+	static bool unsafe_component_dynamic_get_property_list(godex::component_id p_component_id, const void *p_component, List<PropertyInfo> *p_list);
```
Comment (AndreaCatania) [inline ecs.h:258]: This should return void

File: ecs.h [hunk 3]
```diff
@@ -268,6 +274,9 @@ class ECS : public Object {
 	static bool unsafe_databag_set_by_index(godex::databag_id p_databag_id, void *p_databag, uint32_t p_index, const Variant &p_data);
 	static bool unsafe_databag_get_by_index(godex::databag_id p_databag_id, const void *p_databag, uint32_t p_index, Variant &r_data);
 	static void unsafe_databag_call(godex::databag_id p_databag_id, void *p_databag, const StringName &p_method, const Variant **p_args, int p_argcount, Variant *r_ret, Callable::CallError &r_error);
+	static bool unsafe_databag_dynamic_get(godex::databag_id p_databag_id, const void *p_databag, const StringName &p_name, const Variant &p_property);
+	static bool unsafe_databag_dynamic_set(godex::databag_id p_databag_id, const void *p_databag, const StringName &p_name, Variant &r_property);
+	static bool unsafe_databag_dynamic_get_property_list(godex::databag_id p_databag_id, const void *p_databag, List<PropertyInfo> *p_list);
```
Comment (AndreaCatania) [inline ecs.h:279]: This should return void

File: ecs.h [hunk 4]
```diff
@@ -288,6 +297,9 @@ class ECS : public Object {
 	static bool unsafe_event_set_by_index(godex::event_id p_event_id, void *p_event, uint32_t p_index, const Variant &p_data);
 	static bool unsafe_event_get_by_index(godex::event_id p_event_id, const void *p_event, uint32_t p_index, Variant &r_data);
 	static void unsafe_event_call(godex::event_id p_event_id, void *p_event, const StringName &p_method, const Variant **p_args, int p_argcount, Variant *r_ret, Callable::CallError &r_error);
+	static bool unsafe_event_dynamic_get(godex::event_id p_event_id, const void *p_event, const StringName &p_name, const Variant &p_property);
+	static bool unsafe_event_dynamic_set(godex::event_id p_event_id, const void *p_event, const StringName &p_name, Variant &r_property);
+	static bool unsafe_event_dynamic_get_property_list(godex::event_id p_event_id, const void *p_event, List<PropertyInfo> *p_list);
```
Comment (AndreaCatania) [inline ecs.h:302]: This should return void

File: ecs_types.h [hunk 5]
```diff
@@ -188,6 +209,30 @@ public:
...
+	static bool dynamic_set(const void *p_self, const uint32_t p_index, Variant &p_data) {                                             \
+		if constexpr (godex_has__set<m_class>::value) {                                                                                 \
+			const m_class *self = static_cast<const m_class *>(p_self);                                                                \
+			return self->_set(p_index, p_data);                                                                                        \
+		} else {                                                                                                                       \
+			return false;                                                                                                              \
+		}                                                                                                                              \
+	}                                                                                                                                  \
+	static bool dynamic_get_property_list(const void *p_self, List<PropertyInfo> *r_list) {                                            \
```
Comment (AndreaCatania) [inline ecs_types.h:229]: No need to return bool. It's enough don't touch the `r_list`.

File: ecs_types.h [hunk 6]
```diff
@@ -188,6 +209,30 @@ public:
...
+			const m_class *self = static_cast<const m_class *>(p_self);                                                                \
+			return self->_get(p_name, r_data);                                                                                         \
+		} else {                                                                                                                       \
+			return false;                                                                                                              \
+		}                                                                                                                              \
+	}                                                                                                                                  \
+	static bool dynamic_set(void *p_self, const StringName &p_name, const Variant &p_data) {                                           \
+		if constexpr (godex_has__set<m_class>::value) {                                                                                \
+			const m_class *self = static_cast<const m_class *>(p_self);                                                                \
```
Comment (AndreaCatania) [inline ecs_types.h:223]: This should be: `const m_class *self = static_cast<m_class *>(p_self);`

File: ecs_types.h [hunk 7]
```diff
@@ -188,6 +209,28 @@ public:
...
+			const m_class *self = static_cast<m_class *>(p_self);                                                                      \
+			return self->_get(p_name, r_data);                                                                                         \
+		} else {                                                                                                                       \
+			return false;                                                                                                              \
+		}                                                                                                                              \
+	}                                                                                                                                  \
+	static bool dynamic_set(void *p_self, const StringName &p_name, const Variant &p_data) {                                           \
+		if constexpr (godex_has__set<m_class>::value) {                                                                                \
+			const m_class *self = static_cast<m_class *>(p_self);                                                                      \
```
Comment (AndreaCatania) [inline ecs_types.h:223]: The `const m_class *self` should be `m_class *self`

File: ecs_types.h [hunk 8]
```diff
@@ -17,17 +17,38 @@ template <typename T>
...
+struct godex_has__get_property_list<T, decltype(void(std::declval<T &>()._get_property_list(std::declval<List<PropertyInfo> *>())))> : std::true_type {};
 
 template <typename T, typename = void>
 struct godex_has_storage_config : std::false_type {};
 
 template <typename T>
 struct godex_has_storage_config<T, decltype(void(std::declval<T &>()._get_storage_config(std::declval<Dictionary &>())))> : std::true_type {};
 
+template <typename T, typename = void>
+struct godex_has_is_event : std::false_type {};
```
Comment (AndreaCatania) [inline ecs_types.h:50]: I've just removed this code, which was just some left over, in this PR: https://github.com/GodotECS/godex/pull/238. I think you added it back for mistake, during a rebase, you can just remove it.

Comment (RicardRC) [reply ecs_types.h:50]: my bad!

File: ecs_types.h [hunk 9]
```diff
@@ -133,61 +151,110 @@ constexpr SID SID_NONE = UINT32_MAX;
...
+	}                                                                                                                                                     \
+	static bool get_by_index(const void *p_self, const uint32_t p_index, Variant &r_data) {                                                               \
+		const m_class *self = static_cast<const m_class *>(p_self);                                                                                       \
+		ERR_FAIL_COND_V_MSG(p_index >= getters.size(), false, "The parameter " + itos(p_index) + " doesn't exist in this component.");                    \
+		return getters[p_index](self, r_data);                                                                                                            \
+	}                                                                                                                                                     \
+	static bool dynamic_get(void *p_self, const StringName &p_name, Variant &r_data) {                                                                    \
```
Comment (AndreaCatania) [inline ecs_types.h:247]: This is unused now, can you please remove it?

File: ecs_types.h [hunk 10]
```diff
@@ -133,61 +151,110 @@ constexpr SID SID_NONE = UINT32_MAX;
...
+		m_class *self = static_cast<m_class *>(p_self);                                                                                                   \
+		return godex::global_dynamic_get(self, p_name, r_data);                                                                                           \
+	}                                                                                                                                                     \
+	static bool dynamic_set(void *p_self, const StringName &p_name, const Variant &p_data) {                                                              \
```
Comment (AndreaCatania) [inline ecs_types.h:251]: This is unused now, can you please remove it?

**Review by AndreaCatania — CHANGES_REQUESTED** [review]
Everything is perfect now, just some leftover code to remove!

**Review by AndreaCatania — APPROVED** [review]
Perfect!

**Pull request merged**